Migrate from SDL2 to SDL3

rom1v opened the pull request:
First, this PR contains some fixes that could be merged directly into `dev`.

Then, there are some refactors to make the transition easier, especially the handling of integer scrolling (https://github.com/Genymobile/scrcpy/issues/6156#issuecomment-3016836940) and the configuration of the color space at texture creation.

Finally, there's a "big" migration commit, followed by the changes for the build script and the GitHub Actions script.

Links:
 - SDL3 migration docs: https://wiki.libsdl.org/SDL3/README-migration
 - Relevant discussion about the new SDL3 audio API for scrcpy: https://github.com/libsdl-org/SDL/issues/13319
 - Release binaries from GitHub Actions (download `scrcpy-release-sdl3`): https://github.com/rom1v/scrcpy/actions/runs/16226120027

----

Tests and reviews welcome.

Remaining questions:
 - check if the "continuous resizing" workaround is still necessary, or if it can be removed (0b1e59186f1fb75d74f2929f636128958b067ab4) (answer: :x: no https://github.com/Genymobile/scrcpy/pull/6216#issuecomment-3094411968)
 - is it possible with SDL3 to keep the window aspect ratio while resizing (instead of adding black bars)? (#3460)

---

I don't plan to merge this immediately, SDL3 is not available yet on many distros (not even in current Debian stable).

I'd like to implement hardware decoding first (#3800 | https://github.com/libsdl-org/SDL/issues/5405#issuecomment-1754956025 | https://github.com/libsdl-org/SDL/pull/8366), and if everything works as expected, release a new major version (scrcpy 4.0).

EDIT: hardware decoding will not be for 4.0. Flex display was implemented instead (#6772).

rom1v committed:
Handle mouse integer scrolling locally

Do not rely on SDL to expose integer scroll values, as they are not
available in all versions.

It was reimplemented in SDL 3.2.12 by this commit:
<https://github.com/libsdl-org/SDL/commit/0447c2f3c3f3bdf5a302bda50509862903825b5b>

Refs #6156 <https://github.com/Genymobile/scrcpy/issues/6156>
PR #6216 <https://github.com/Genymobile/scrcpy/pull/6216>

rom1v committed:
Remove prepare_for_frame()

Inline the function body into its only caller to simplify further
refactors.

PR #6216 <https://github.com/Genymobile/scrcpy/pull/6216>

rom1v committed:
Create texture on first frame

The texture was created as soon as the initial video size was known,
even before the first frame arrived.

However, texture creation will require other data, such as the color
range, which is only available once the first frame is received.

Therefore, delay texture creation until the first frame.

PR #6216 <https://github.com/Genymobile/scrcpy/pull/6216>

rom1v committed:
Set color range during texture creation

This prepares for the migration to SDL3, where the color range can only
be specified at the time of texture creation.

PR #6216 <https://github.com/Genymobile/scrcpy/pull/6216>

rom1v committed:
Improve color space conversion

Use both the color space and color range from FFmpeg to determine the
appropriate SDL YUV conversion mode.

PR #6216 <https://github.com/Genymobile/scrcpy/pull/6216>

rom1v committed:
Migrate from SDL2 to SDL3

Refs <https://wiki.libsdl.org/SDL3/README-migration>
PR #6216 <https://github.com/Genymobile/scrcpy/pull/6216>

rom1v committed:
Build SDL3 for test step on Github Actions

The Ubuntu version used to build scrcpy does not provide an SDL3
package.

PR #6216 <https://github.com/Genymobile/scrcpy/pull/6216>

yume-chan commented:
> How much data does it require:

`additional_amount` is always `3840`, and audio sounds like a robot

> Is it different with `--audio-output-buffer=10` or `--audio-output-buffer=20`? 

There is no error with either `--audio-output-buffer=10` or `--audio-output-buffer=20`, and audio is working

> On SDL2, do you need to use --audio-output-buffer=xx to get audio working correctly?

No

rom1v commented:
> There is no error with either `--audio-output-buffer=10` or `--audio-output-buffer=20`, and audio is working

What is the value of `additional_amount` in these cases?

@icculus Sorry to ping you here, but do you know why setting `SDL_HINT_AUDIO_DEVICE_SAMPLE_FRAMES` to 240 (5 milliseconds of samples, 1920 bytes for 2 channels of 4 bytes each) can result in the callback being called with `additional_amount = 3840` (instead of 1920) on Windows?

Apparently, on SDL2, the callback was correctly called with 1920 bytes. @yume-chan, this can be confirmed on scrcpy 3.3.1 with:

```diff
diff --git app/src/audio_player.c app/src/audio_player.c
index 9413c2ea8..07979b31b 100644
--- app/src/audio_player.c
+++ app/src/audio_player.c
@@ -11,6 +11,8 @@ static void SDLCALL
 sc_audio_player_sdl_callback(void *userdata, uint8_t *stream, int len_int) {
     struct sc_audio_player *ap = userdata;
 
+    LOGI("=== SDL2 audio callback: %d bytes (%d samples)", len_int, len_int / 8);
+
     assert(len_int > 0);
     size_t len = len_int;
 
```

metayan commented on app/deps/sdl.sh:
```diff
@@ -35,45 +35,48 @@ else
         conf+=(
-            --enable-video-wayland
-            --enable-video-x11
+            -DSDL_WAYLOAD=ON
```
You probably mean SDL_WAYLAND?

rom1v replied:
Good catch :+1:

Indeed, in the logs:

```
 CMake Warning:
-- Generating done (0.1s)
  Manually-specified variables were not used by the project:

    SDL_WAYLOAD
```

(it should be an error instead of a warning)

metayan commented:
```diff
diff --git a/app/deps/sdl.sh b/app/deps/sdl.sh
index ae79ae9c..a308ae9e 100755
--- a/app/deps/sdl.sh
+++ b/app/deps/sdl.sh
@@ -28,7 +28,7 @@ export CXXFLAGS="$CFLAGS"
 
 if [[ -d "$DIRNAME" ]]
 then
-    echo "'$PWD/$HDIRNAME' already exists, not reconfigured"
+    echo "'$PWD/$DIRNAME' already exists, not reconfigured"
     cd "$DIRNAME"
 else
     mkdir "$DIRNAME"
@@ -41,7 +41,7 @@ else
     if [[ "$HOST" == linux ]]
     then
         conf+=(
-            -DSDL_WAYLOAD=ON
+            -DSDL_WAYLAND=ON
             -DSDL_X11=ON
         )
     fi
```

metayan commented:
Also tested with SDL 3.2.18 and it works fine on macOS. Audio not tested.
```diff
diff --git a/app/deps/sdl.sh b/app/deps/sdl.sh
index ae79ae9c..970fcbfa 100755
--- a/app/deps/sdl.sh
+++ b/app/deps/sdl.sh
@@ -5,10 +5,10 @@ cd "$DEPS_DIR"
 . common
 process_args "$@"
 
-VERSION=3.2.16
+VERSION=3.2.18
 FILENAME=SDL-$VERSION.tar.gz
 PROJECT_DIR=SDL-release-$VERSION
-SHA256SUM=a1af401fb49b824ba721683862439e014498f79753602827e174f3613357aff0
+SHA256SUM=51539fa13e546bc50c632beed3f34257de2baa38a4c642048de56377903b4265
 
 cd "$SOURCES_DIR"
```

yume-chan commented:
> What is the value of `additional_amount` in these cases?

It's still `3840`

```diff
diff --git a/app/src/audio_player.c b/app/src/audio_player.c
index e1a4134e..71fbef2a 100644
--- a/app/src/audio_player.c
+++ b/app/src/audio_player.c
@@ -16,6 +16,8 @@ sc_audio_player_stream_callback(void *userdata, SDL_AudioStream *stream,
     struct sc_audio_player *ap = userdata;
 
     if (additional_amount > 0) {
+        LOGE("additional_amount: %d", additional_amount);
+
         size_t len = additional_amount;
 
         assert(len <= ap->aout_buffer_size);
@@ -82,6 +84,9 @@ sc_audio_player_frame_sink_open(struct sc_frame_sink *sink,
     int r = snprintf(str, sizeof(str), "%" PRIu16, (uint16_t) aout_samples);
     assert(r >= 0 && (size_t) r < sizeof(str));
     (void) r;
+
+    LOGE("setting SDL_HINT_AUDIO_DEVICE_SAMPLE_FRAMES to %s", str);
+
     if (!SDL_SetHint(SDL_HINT_AUDIO_DEVICE_SAMPLE_FRAMES, str)) {
         LOGE("Could not set audio output buffer");
         sc_audio_regulator_destroy(&ap->audioreg);
```

```txt
> ./scrcpy --audio-output-buffer=10
scrcpy 3.3.1 <https://github.com/Genymobile/scrcpy>
INFO: ADB device found:
INFO:     -->   (usb)  73294bba                        device  M2011K2C
D:\dev\yume-chan\scrcpy-devcontainer\scrcpy\server\build\o...file pushed, 0 skipped. 102.8 MB/s (91004 bytes in 0.001s)
[server] INFO: Device: [Xiaomi] Xiaomi M2011K2C (Android 13)
INFO: Renderer: direct3d11
ERROR: setting SDL_HINT_AUDIO_DEVICE_SAMPLE_FRAMES to 480
ERROR: additional_amount: 3840
ERROR: additional_amount: 3840
ERROR: additional_amount: 3840
ERROR: additional_amount: 3840
ERROR: additional_amount: 3840
ERROR: additional_amount: 3840
```

```txt
> ./scrcpy --audio-output-buffer=20
scrcpy 3.3.1 <https://github.com/Genymobile/scrcpy>
INFO: ADB device found:
INFO:     -->   (usb)  73294bba                        device  M2011K2C
D:\dev\yume-chan\scrcpy-devcontainer\scrcpy\server\build\o...file pushed, 0 skipped. 107.7 MB/s (91004 bytes in 0.001s)
[server] INFO: Device: [Xiaomi] Xiaomi M2011K2C (Android 13)
INFO: Renderer: direct3d11
ERROR: setting SDL_HINT_AUDIO_DEVICE_SAMPLE_FRAMES to 960
ERROR: additional_amount: 3840
ERROR: additional_amount: 3840
ERROR: additional_amount: 3840
INFO: Texture: 1440x3200
ERROR: additional_amount: 3840
ERROR: additional_amount: 3840
```

---

> this can be confirmed on scrcpy 3.3.1

```txt
> ./scrcpy
scrcpy 3.3.1 <https://github.com/Genymobile/scrcpy>
INFO: ADB device found:
INFO:     -->   (usb)  73294bba                        device  M2011K2C
D:\dev\yume-chan\scrcpy-devcontainer\scrcpy\server\build\o... file pushed, 0 skipped. 99.5 MB/s (91004 bytes in 0.001s)
[server] INFO: Device: [Xiaomi] Xiaomi M2011K2C (Android 13)
INFO: Renderer: direct3d
INFO: Texture: 1440x3200
ERROR: === SDL2 audio callback: 1920 bytes (240 samples)
ERROR: === SDL2 audio callback: 1920 bytes (240 samples)
ERROR: === SDL2 audio callback: 1920 bytes (240 samples)
ERROR: === SDL2 audio callback: 1920 bytes (240 samples)
ERROR: === SDL2 audio callback: 1920 bytes (240 samples)
```

```txt
> ./scrcpy --audio-output-buffer=10
scrcpy 3.3.1 <https://github.com/Genymobile/scrcpy>
INFO: ADB device found:
INFO:     -->   (usb)  73294bba                        device  M2011K2C
D:\dev\yume-chan\scrcpy-devcontainer\scrcpy\server\build\o...file pushed, 0 skipped. 104.3 MB/s (91004 bytes in 0.001s)
[server] INFO: Device: [Xiaomi] Xiaomi M2011K2C (Android 13)
INFO: Renderer: direct3d
INFO: Texture: 1440x3200
ERROR: === SDL2 audio callback: 3840 bytes (480 samples)
ERROR: === SDL2 audio callback: 3840 bytes (480 samples)
ERROR: === SDL2 audio callback: 3840 bytes (480 samples)
ERROR: === SDL2 audio callback: 3840 bytes (480 samples)
ERROR: === SDL2 audio callback: 3840 bytes (480 samples)
ERROR: === SDL2 audio callback: 3840 bytes (480 samples)
```

```txt
> ./scrcpy --audio-output-buffer=20
scrcpy 3.3.1 <https://github.com/Genymobile/scrcpy>
INFO: ADB device found:
INFO:     -->   (usb)  73294bba                        device  M2011K2C
D:\dev\yume-chan\scrcpy-devcontainer\scrcpy\server\build\o...file pushed, 0 skipped. 103.7 MB/s (91004 bytes in 0.001s)
[server] INFO: Device: [Xiaomi] Xiaomi M2011K2C (Android 13)
INFO: Renderer: direct3d
INFO: Texture: 1440x3200
ERROR: === SDL2 audio callback: 7680 bytes (960 samples)
ERROR: === SDL2 audio callback: 7680 bytes (960 samples)
ERROR: === SDL2 audio callback: 7680 bytes (960 samples)
ERROR: === SDL2 audio callback: 7680 bytes (960 samples)
ERROR: === SDL2 audio callback: 7680 bytes (960 samples)
```

metayan commented:
Numbers and punctuation now only work with `--raw-key-events`.

yume-chan commented:
https://github.com/libsdl-org/SDL/blob/bc5c9a686ca01c7cffa45e223efcd0cb7ef1efb7/src/audio/wasapi/SDL_wasapi.c#L759-L760

SDL3 calls [IAudioClient3::GetSharedModeEnginePeriod](https://learn.microsoft.com/en-us/windows/win32/api/audioclient/nf-audioclient-iaudioclient3-getsharedmodeengineperiod) and clamps `device->sample_frames`. `min_period_in_frames` and `max_period_in_frames` are both `480`, so `additional_amount` is always `3840`. Not sure what it means, nor how does SDL2 works.

rom1v commented:
> Numbers and punctuation now only work with `--raw-key-events`.

Oh, `SDL_EVENT_TEXT_INPUT` is never received: https://github.com/Genymobile/scrcpy/blob/3615c75ad6074f088cddb0616410022ce7c0249b/app/src/input_manager.c#L1019

rom1v commented:
> `directsound` audio driver doesn't work with the default `--audio-output-buffer=5` neither, `additional_amount` is 1920 as expected, but audio played is broken

Is it also the case on SDL2?

`--audio-output-buffer` was added because some audio outputs on Windows did not support 5ms, so if the behavior was the same on SDL2 (and respects the `sample_frames` value), it's ok I guess: https://github.com/Genymobile/scrcpy/issues/3793

rom1v commented:
> Oh, SDL_EVENT_TEXT_INPUT is never received

Should be fixed on the new branch.

I added:

```diff
diff --git app/src/screen.c app/src/screen.c
index 27259e830..6ecb64167 100644
--- app/src/screen.c
+++ app/src/screen.c
@@ -394,6 +394,12 @@ sc_screen_init(struct sc_screen *screen,
         goto error_destroy_fps_counter;
     }
 
+    ok = SDL_StartTextInput(screen->window);
+    if (!ok) {
+        LOGE("Could not enable text input: %s", SDL_GetError());
+        goto error_destroy_window;
+    }
+
     SDL_Surface *icon = scrcpy_icon_load();
     if (icon) {
         SDL_SetWindowIcon(screen->window, icon);
```

(cf https://github.com/libsdl-org/SDL/issues/13380)

As a consequence, it might open an IME on scrcpy start (https://github.com/libsdl-org/SDL/issues/13380#issuecomment-3083668746). If that's the case (to be tested), we need a workaround.

yume-chan commented:
> > `directsound` audio driver doesn't work with the default `--audio-output-buffer=5` neither, `additional_amount` is 1920 as expected, but audio played is broken
> 
> Is it also the case on SDL2?

Yes, with SDL2, directsound driver, and `--audio-output-buffer=5`, audio playback is also broken. `wasapi` driver works.

rom1v commented:
Thank you :+1:

Maybe it can make sense to use SDL main callbacks (to be tested). This will require to rework the way the scrcpy-server is started, but in the end in might simplify (or not).

rom1v commented:
@yume-chan

> === SDL2 audio callback: 1920 bytes (240 samples)

Could you please also add the timestamp of the call, to see if the callback is called twice in a row: https://github.com/libsdl-org/SDL/issues/13397

adamponi commented:
I can't reproduce this using SDL2 because the version built on github doesn't offer a Vulkan renderer.

adamponi commented:
`PS C:\Users\xxxxxxxxxxxx\Downloads\scrcpy-release-sdl3\scrcpy-win64-sdl3> ./scrcpy --render=vulkan --no-audio -m1024
scrcpy 3.3.1 <https://github.com/Genymobile/scrcpy>
INFO: ADB device found:
INFO:     -->   (usb)  ec8faaa3                        device  xxxxxxxxxxxx
C:\Users\xxxxxxxxxxxx\Downloads\scrcpy-release-sdl3\scrcpy-... file pushed, 0 skipped. 28.7 MB/s (90788 bytes in 0.003s)
[server] INFO: Device: [Xiaomi] Redmi xxxxxxxxxxxx (Android 13)
INFO: Renderer: vulkan
INFO: Texture: 464x1024
`
<img width="471" height="950" alt="image" src="https://github.com/user-attachments/assets/253b92df-94d5-4711-a83a-f79dd9c3560e" />
This parameter didn't help. The problem is the same. I tested several other renderers, and the ones I tested were:
-direct3d12
-direct3d11
-opengl

The problem doesn't occur. The problem only occurs with the Vulkan renderer.

metayan commented:
Paste doesn't work on macOS.
Tried SDL 3.2.16 3.2.18 3.2.20.

SDL issue filed:
https://github.com/libsdl-org/SDL/issues/13737

Copy works and sometimes Control-Shift-V works.

dddddjent committed:
Fix SDL colorspace matching AVCOL_SPC_RGB

Use BT.709 color space for AVCOL_SPC_RGB.

Refs #1868 comment <https://github.com/Genymobile/scrcpy/issues/1868#issuecomment-3293917796>

Signed-off-by: Romain Vimont <rom@rom1v.com>

rom1v committed:
Remove obsolete workaround for macOS

Refs #3031 <https://github.com/Genymobile/scrcpy/pull/3031>
Refs SDL/#5340 <https://github.com/libsdl-org/SDL/issues/5340>
Refs SDL/#5976 <https://github.com/libsdl-org/SDL/issues/5976>
PR #6216 <https://github.com/Genymobile/scrcpy/pull/6216>

rom1v committed:
Check programming errors reported by SDL3

Refs #6216 comment <https://github.com/Genymobile/scrcpy/pull/6216#issuecomment-3399353701>
Refs SDL/#14223 <https://github.com/libsdl-org/SDL/issues/14223>
PR #6216 <https://github.com/Genymobile/scrcpy/pull/6216>

rom1v commented:
I added some changes on this branch.

Firstly, I removed a workaround for macOS (e59d40ebc7507a2eee75a1ea6c350768b4f713b5), because the underlying issue should be fixed in SDL3 (https://github.com/libsdl-org/SDL/issues/5340, https://github.com/libsdl-org/SDL/issues/5976). Please verify on macOS.

Then, I noticed a very annoying change from SDL2 to SDL3: many previously infallible functions have been made fallible without good reason (in my opinion). See the discussion in https://github.com/libsdl-org/SDL/issues/14223 for details.

For example, `SDL_GetWindowSize()`, which returned `void` in SDL2, now returns `bool` in SDL3. In general, it's a good thing that a function reports an error when an error can occur. However, in these changes, the functions only report programming errors, basically when SDL is not initialized or an invalid argument is passed (at least for cases they can detect):

https://github.com/libsdl-org/SDL/blob/5dab2c73f0537bd121d2311b6949a654c5692af9/src/video/SDL_video.c#L3203-L3213
https://github.com/libsdl-org/SDL/blob/5dab2c73f0537bd121d2311b6949a654c5692af9/src/video/SDL_video.c#L159-L178

Handling such errors would have cascading consequences in scrcpy. For instance, if a function calls `SDL_GetWindowSize()`, it should report the error to its caller, which would also have be made fallible, and so on… adding a lot of complexity for no reason, since in practice, `SDL_GetWindowSize()` cannot fail except due to programming errors.

The SDL maintainers disagree with the statement that "these functions only fails if the object is invalid", and suggest that these functions may fail for other reasons in the future.

In practice, I still assume that they may not fail, and that they never will (at least for `SDL_GetWindowXXX()`):
 - The current implementation cannot fail when the functions are called correctly.
 - If they add genuine failure cases in the future, every call to `SDL_GetWindowSize()` that doesn't check the error in a project would effectively be a bug (leading to undefined behavior if the output variables are not initialized beforehand).
 - The migration script from SDL2 to SDL3 does not add these error checks, so this would basically introduce undefined behavior in every migrated project.

Therefore, I added a wrapper in scrcpy that check the reported values, log them when failing is not fatal, and `abort()` (think "assertion enabled in release mode") to avoid undefined behavior for window size and position getters: a9a8fb190cb8dd9b2db1021c9bf0fbe3cf2a78f4.

Error handling could be added later if my assumptions prove to be incorrect.

What do you think?

EDIT: see #6802

rom1v committed:
Upgrade SDL build script for SDL3

PR #6216 <https://github.com/Genymobile/scrcpy/pull/6216>

anotheruserofgithub commented on app/src/icon.c:
```diff
@@ -248,17 +236,35 @@ load_from_path(const char *path) {
+    // frame owns the data
+    bool ok = SDL_SetPointerProperty(props, "sc_frame", frame);
+    if (!ok) {
+        LOGE("Could not get surface properties: %s", SDL_GetError());
```
```suggestion
        LOGE("Could not set pointer property: %s", SDL_GetError());
```

anotheruserofgithub commented on app/src/input_manager.c:
```diff
@@ -911,31 +906,31 @@ sc_input_manager_process_mouse_wheel(struct sc_input_manager *im,
-            SDL_GameControllerClose(gc);
+            SDL_CloseGamepad(gc);
             return;
         }
```
Same things as in previous comments: reword "game controller" and "controller" to "gamepad" and find a better name for the `gc` pointer.

anotheruserofgithub reviewed (commented):
Hi, thanks for your great work! Here are a few things I spotted in commit c7a4717d9862e3cd2f4bd38c25341eb4b442c79b (the others are neat to me), though I didn't check the lines that aren't modified.

Unfortunately I don't have any device to test. :( Sorry! Only got a very old phone and USB debugging isn't detected apparently (the USB socket is in a very bad state and is barely capable of charging the battery, might be the issue).

anotheruserofgithub reviewed (commented):
Comments, especially in file `input_events.h` (but also in other files, see https://github.com/Genymobile/scrcpy/pull/6216#discussion_r2651778910), should be reviewed too.

I'm referring to lines in [`app/src/input_events.h`](https://github.com/Genymobile/scrcpy/blob/db2f72a58aaedfe3384b0dea838175e9237a3324/app/src/input_events.h) on the current master branch (commit db2f72a58aaedfe3384b0dea838175e9237a3324):
- L17 there is a `sc_gamepad_processor` too.
- L326 the migration guide has been moved to https://wiki.libsdl.org/SDL3/README-migration (hyphen instead of slash). But maybe you should now point to the actual documentation / wiki page? No longer need to refer to the SDL2 -> SDL3 transition.
- L416 wrongly points to SDL2 wiki.
- L505 should be `SDL_GAMEPAD_AXIS_*`.
- L515 should be `SDL_GAMEPAD_BUTTON_*`.

Additionally in [`app/src/mouse_capture.c`](https://github.com/Genymobile/scrcpy/blob/db2f72a58aaedfe3384b0dea838175e9237a3324/app/src/mouse_capture.c#L63):
- L63 the event new name is `SDL_EVENT_MOUSE_BUTTON_UP`.

Also links to SDL2 header files should be changed to SDL3, here in [`app/src/hid/hid_keyboard.h`](https://github.com/Genymobile/scrcpy/blob/db2f72a58aaedfe3384b0dea838175e9237a3324/app/src/hid/hid_keyboard.h#L12): https://github.com/Genymobile/scrcpy/blob/db2f72a58aaedfe3384b0dea838175e9237a3324/app/src/hid/hid_keyboard.h#L12

And library versions should be adapted (**two** occurrences in each block): https://github.com/Genymobile/scrcpy/blob/db2f72a58aaedfe3384b0dea838175e9237a3324/doc/linux.md?plain=1#L42-L43 https://github.com/Genymobile/scrcpy/blob/db2f72a58aaedfe3384b0dea838175e9237a3324/.github/workflows/release.yml#L78-L79 https://github.com/Genymobile/scrcpy/blob/db2f72a58aaedfe3384b0dea838175e9237a3324/.github/workflows/release.yml#L104-L105 https://github.com/Genymobile/scrcpy/blob/db2f72a58aaedfe3384b0dea838175e9237a3324/.github/workflows/release.yml#L135-L136 https://github.com/Genymobile/scrcpy/blob/db2f72a58aaedfe3384b0dea838175e9237a3324/.github/workflows/release.yml#L166-L167

I may have missed other details.

rom1v commented:
@anotheruserofgithub Thank you very much for your review. :+1:

I fixed most of them. I will investigate the remaining ones (especially https://github.com/Genymobile/scrcpy/pull/6216#discussion_r2651810005).

> L17 there is a `sc_gamepad_processor` too.

I added a separate commit.

> L326 the migration guide has been moved to https://wiki.libsdl.org/SDL3/README-migration (hyphen instead of slash). But maybe you should now point to the actual documentation / wiki page? No longer need to refer to the SDL2 -> SDL3 transition.

I removed the comment, it was only necessary when the named differred from SDL.

> (changed in `scrcpy/.github/workflows/release.yml`)

These ones must still refer to SDL2:
 - the old Ubuntu version used does not package SDL3
 - this is necessary (at least, this is an easy way) to install the SDL dependencies required to build a working static binary

anotheruserofgithub commented on app/src/input_manager.c:
```diff
@@ -911,31 +906,31 @@ sc_input_manager_process_mouse_wheel(struct sc_input_manager *im,
+    if (event->type == SDL_EVENT_GAMEPAD_ADDED) {
+        SDL_Gamepad *sdl_gamepad = SDL_OpenGamepad(event->which);
+        if (!sdl_gamepad) {
             LOGW("Could not open game controller");
```
```suggestion
            LOGW("Could not open gamepad");
```

anotheruserofgithub commented on app/src/input_manager.c:
```diff
@@ -911,31 +906,31 @@ sc_input_manager_process_mouse_wheel(struct sc_input_manager *im,
-        SDL_Joystick *joystick = SDL_GameControllerGetJoystick(gc);
+        SDL_Joystick *joystick = SDL_GetGamepadJoystick(sdl_gamepad);
         if (!joystick) {
             LOGW("Could not get controller joystick");
```
```suggestion
            LOGW("Could not get gamepad joystick");
```

anotheruserofgithub reviewed (commented):
Leftovers from https://github.com/Genymobile/scrcpy/pull/6216#discussion_r2651850473.

anotheruserofgithub reviewed (commented):
Typo when fixing https://github.com/Genymobile/scrcpy/pull/6216#pullrequestreview-3616928335.

anotheruserofgithub commented on app/src/main.c:
```diff
@@ -5,8 +5,9 @@
-#define SDL_MAIN_HANDLED // avoid link error on Linux Windows Subsystem
-#include <SDL2/SDL.h>
+#define SDL_FUNCTION_POINTER_IS_VOID_POINTER
+#include <SDL3/SDL_stdinc.h>
```
```suggestion
```

I scratched my head when reading https://wiki.libsdl.org/SDL3/README-main-functions#the-main-entry-point-in-sdl3 but now I think I understand the global picture.

In SDL2:
- `SDL.h` includes `SDL_main.h` [here](https://github.com/libsdl-org/SDL/blob/SDL2/include/SDL.h#L31).
- `SDL_main.h` checks for a defined symbol `SDL_MAIN_HANDLED` to avoid redefining the `main` function [here](https://github.com/libsdl-org/SDL/blob/SDL2/include/SDL_main.h#L33).
- Thus scrcpy has to define `SDL_MAIN_HANDLED` because it includes `SDL.h` which includes `SDL_main.h`.
- This is important when it is linked to the `SDLmain` static library which provides the real `main` entry point.
- `SDL.h` includes `SDL_stdinc.h` [here](https://github.com/libsdl-org/SDL/blob/SDL2/include/SDL.h#L32).
- `SDL_stdinc.h` does not check `SDL_FUNCTION_POINTER_IS_VOID_POINTER` nor does it define `SDL_FunctionPointer`.

In SDL3:
- `SDL.h` does *not* include `SDL_main.h` [here](https://github.com/libsdl-org/SDL/blob/main/include/SDL3/SDL.h#L33).
- `SDL_main.h` checks for a defined symbol `SDL_MAIN_HANDLED` to avoid redefining the `main` function [here](https://github.com/libsdl-org/SDL/blob/main/include/SDL3/SDL_main.h#L147).
- Thus scrcpy does not need to define `SDL_MAIN_HANDLED` because it includes `SDL.h` but which does not include `SDL_main.h`.
- This is on purpose since the `SDLmain` static library has been replaced by a single-header library (`SDL_main.h` does the whole work now).
- `SDL.h` includes `SDL_stdinc.h` [here](https://github.com/libsdl-org/SDL/blob/main/include/SDL3/SDL.h#L34).
- `SDL_stdinc.h` checks `SDL_FUNCTION_POINTER_IS_VOID_POINTER` [here](https://github.com/libsdl-org/SDL/blob/main/include/SDL3/SDL_stdinc.h#L6156) in order to define `SDL_FunctionPointer`.
- Hence scrcpy has to define `SDL_FUNCTION_POINTER_IS_VOID_POINTER` so as to work around a compiler error somewhere (right?).
- However scrcpy does not need to explicitly include `SDL_stdinc.h` because it is already included by `SDL.h`.

anotheruserofgithub reviewed (commented):
As a consequence I expect you can also remove `SDL_MAIN_HANDLED` for tests in `app/meson.build` (see commit ba1b36758e8a6cbebe8215d9bb0936d4598d6045): https://github.com/Genymobile/scrcpy/blob/db2f72a58aaedfe3384b0dea838175e9237a3324/app/meson.build#L278

But do they need `SDL_FUNCTION_POINTER_IS_VOID_POINTER`?

anotheruserofgithub reviewed (commented):
An occurrence of "controller" that I missed.

The last one is in `README.md`: https://github.com/Genymobile/scrcpy/blob/db2f72a58aaedfe3384b0dea838175e9237a3324/README.md?plain=1#L132 Since there is already the verb "Control" perhaps you want to change `gamepad controllers` to `gamepads`? (More generically these are "game controllers", not "gamepad controllers".)

rom1v commented on app/src/screen.c:
```diff
@@ -712,8 +712,8 @@ void
+    if (!ok) {
         LOGW("Could not switch fullscreen mode: %s", SDL_GetError());
         return;
     }
```
I fixed it in a separate commit for clarity: e24018b4b7fdc3e18eb40d0f958916536358c989

I completely removed the local fields. Even if they were only updated on SDL events, their state was fundamentally inconistent. For example:
 1. start in windowed mode
 2. maximize the window → MAXIMIZED event
 3. enable fullscreen → RESTORED event then ENTER_FULLSCREEN event
 4. disable fullscreen → LEAVE_FULLSCREEN event then MAXIMIZED event

If the `maximized` field was updated on `MAXIMIZED`, then its value would be `false` in `LEAVE_FULLSCREEN`. But `SDL_GetWindowFlags()` still returns a `MAXIMIZED` state (the internal state is set before the event is posted, but the boolean field can only be updated when the event is handled).

So far, I only tested on Linux+X11. The relevant use case is:
 1. start in windowed mode
 2. maximize the window
 3. enable fullscreen
 4. rotate the device (switch portrait/landscape)
 5. disable fullscreen
 6. unmaximize

The resulting window should be resized to the new rotated dimensions.

anotheruserofgithub reviewed (commented):
A few questions about the new commit e24018b4b7fdc3e18eb40d0f958916536358c989.

rom1v committed:
Disable "resize to pixel-perfect" when maximized

This improves consistency and will simplify further refactors.

PR #6216 <https://github.com/Genymobile/scrcpy/pull/6216>

rom1v committed:
Fix asynchronous SDL window state handling

The SDL3 migration guide [1] says:

> The following window operations are now considered to be asynchronous
requests and should not be assumed to succeed unless a corresponding
event has been received […]

Remove the boolean fields used to track the window state, as they were
fundamentally inconsistent with SDL's internal state. Use
SDL_GetWindowFlags() explicitly instead.

[1]: <https://wiki.libsdl.org/SDL3/README-migration>

Refs #6216 comment <https://github.com/Genymobile/scrcpy/pull/6216#discussion_r2651810005>
PR #6216 <https://github.com/Genymobile/scrcpy/pull/6216>

rom1v commented on app/src/screen.c:
```diff
@@ -799,26 +785,18 @@ sc_screen_handle_event(struct sc_screen *screen, const SDL_Event *event) {
                 apply_pending_resize(screen);
-                sc_screen_render(screen, true);
             }
             return true;
```
> On `RESTORED` you resize only if it is a video window, but on `LEAVE_FULLSCREEN` you resize even if there is no video, is it because you allow fullscreen only if `assert(screen->video)`?

Yes, exactly, without video (`--no-video`), there is no way to enable fullscreen. I just added some asserts.

> You removed `sc_screen_render(screen, true)`, is it because it will be done on `PIXEL_SIZE_CHANGED` after `sc_sdl_set_window_size()` is called?

Yes, that was the reason. But I finally reverted that changed, because it seems it's important in some cases (see this old commit 7343b233e457076d9419fb0f6e587ba13a7e79c8). And the change should not belong to this refactor anyway.

anotheruserofgithub replied:
I agree with the latest changes. ;)

rom1v commented on app/src/screen.c:
```diff
@@ -759,15 +750,10 @@ void
-    if (screen->maximized) {
-        sc_sdl_restore_window(screen->window);
-        screen->maximized = false;
-    }
```
Done.

rom1v commented:
You can test the latest version of this PR (branch `sdl3.36`) built here: https://github.com/rom1v/scrcpy/actions/runs/20615612335

rom1v committed:
Upgrade SDL3 dependency list for Github Actions

Explicitly install required packages listed here:
<https://wiki.libsdl.org/SDL3/README-linux#build-dependencies>

Refs #6216 comment <https://github.com/Genymobile/scrcpy/pull/6216#issuecomment-3703680127>
PR #6216 <https://github.com/Genymobile/scrcpy/pull/6216>

anotheruserofgithub commented:
> In fact, the list of dependencies is specified here: https://wiki.libsdl.org/SDL3/README-linux#build-dependencies
> 
> So I added commit [33bf758](https://github.com/Genymobile/scrcpy/commit/33bf7587e4c27dc7b0675a5767267dc178742f82).

To cross-build for Windows (32 & 64) do you have to keep installing libsdl2 or can you also install each dependency?

rom1v commented:
> To cross-build for Windows (32 & 64) do you have to keep installing libsdl2 or can you also install each dependency?

We don't need the package at all. I added 1c6ffd4085c0ff567c3ea7fe2eb3102d59101586.

rom1v commented:
This PR is ready to be merged. :rocket: 

Here are binaries for the latest version: https://github.com/rom1v/scrcpy/actions/runs/20693631961

anotheruserofgithub commented on doc/build.md:
There are still some references to SDL2 in this file: https://github.com/Genymobile/scrcpy/blob/e3fa95387483635182962fc1fbdda207bfea5cb7/doc/build.md?plain=1#L33 https://github.com/Genymobile/scrcpy/blob/e3fa95387483635182962fc1fbdda207bfea5cb7/doc/build.md?plain=1#L39 https://github.com/Genymobile/scrcpy/blob/e3fa95387483635182962fc1fbdda207bfea5cb7/doc/build.md?plain=1#L80 https://github.com/Genymobile/scrcpy/blob/e3fa95387483635182962fc1fbdda207bfea5cb7/doc/build.md?plain=1#L124 https://github.com/Genymobile/scrcpy/blob/e3fa95387483635182962fc1fbdda207bfea5cb7/doc/build.md?plain=1#L139

rom1v replied:
Thank you, fixed.

rom1v commented:
:rocket:
